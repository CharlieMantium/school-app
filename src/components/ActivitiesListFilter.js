import React from 'react';
import Input from './Input';

const ActivitiesListFilter = () => (
  <div>
    <Input type="text" />
  </div>
);

export default ActivitiesListFilter;
