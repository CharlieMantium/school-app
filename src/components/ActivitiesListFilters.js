import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {setTextFilter} from '../actions/filters';

const ActivitiesListFilters = ({ filters: {text}, dispatch }) => (
  <div>
    <input type="text" value={text} onChange={(e) => {
      dispatch(setTextFilter(e.target.value));
    }}/>
  </div>
);

ActivitiesListFilters.propTypes = {
  filters: PropTypes.object,
  text: PropTypes.string,
  dispatch: PropTypes.func
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

export default connect(mapStateToProps)(ActivitiesListFilters);
