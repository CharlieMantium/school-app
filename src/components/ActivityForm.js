import React from 'react';

export default class ActivityForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      classNo: props.activity ? props.activity.classNo : '',
      day: props.activity ? props.activity.day : '',
      name: props.activity ? props.activity.name : '',
      room: props.activity ? props.activity.room : '',
      teacher: props.activity ? props.activity.teacher : '',
      error: ''
    };
  };
  onNameChange = (e) => {
    const name = e.target.value;
    this.setState(() => ({name}));
  };
  onTeacherChange = (e) => {
    const teacher = e.target.value;
    this.setState(() => ({teacher}));
  };
  onRoomChange = (e) => {
    const room = e.target.value;
    this.setState(() => ({room}));
  };
  onClassNoChange = (e) => {
    const classNo = e.target.value;
    if (!classNo || classNo.match(/^[1-9]\d*$/)) {
      this.setState(() => ({classNo}));
    };
  };
  onDayChange = (e) => {
    const day = e.target.value;
    this.setState(() => ({day}));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.name || !this.state.day || !this.state.classNo) {
      this.setState(() => ({error: 'Please provide Activity Name, Day and Class Number!'}))
    } else {
      this.setState(() => ({error: ''}))
      this.props.onSubmit({
        classNo: this.state.classNo,
        day: this.state.day,
        name: this.state.name,
        room: this.state.room,
        teacher: this.state.teacher,
      })
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Activity Name"
            autoFocus
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <input
            type="text"
            placeholder="Teacher"
            value={this.state.teacher}
            onChange={this.onTeacherChange}
          />
          <input
            type="text"
            placeholder="Room"
            value={this.state.room}
            onChange={this.onRoomChange}
          />
          <input
            type="number"
            placeholder="Class Number"
            value={this.state.classNo}
            onChange={this.onClassNoChange}
          />
          <select
            value={this.state.day}
            onChange={this.onDayChange}
            required={true}
          >
            <option value=''>Pick a day</option>
            <option value={1}>Monday</option>
            <option value={2}>Tuesday</option>
            <option value={3}>Wednesday</option>
            <option value={4}>Thursday</option>
            <option value={5}>Friday</option>
            <option value={6}>Saturday</option>
            <option value={7}>Sunday</option>
          </select>
          <button>Add Activity</button>
        </form>
      </div>
    )
  }
}
