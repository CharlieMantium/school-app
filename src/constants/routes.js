export const ACTIVITY_PLAN_ROUTE = '/';
export const CREATE_ACTIVITY_ROUTE = '/create';
export const EDIT_ACTIVITY_ID_ROUTE = '/edit/:id';
export const EDIT_ACTIVITY_ROUTE = id => `/edit/${id}`;
