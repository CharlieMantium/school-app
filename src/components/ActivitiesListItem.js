import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { EDIT_ACTIVITY_ROUTE } from '../constants/routes';

const ActivitiesListItem = ({ name, day, classNo, teacher, id }) => (
  <div>
    <Link to={EDIT_ACTIVITY_ROUTE(id)}>
      <h3>{name}</h3>
    </Link>
    <p>{`On ${day} on ${classNo} class with ${teacher}`}</p>
  </div>
);

ActivitiesListItem.propTypes = {
  name: PropTypes.string,
  day: PropTypes.string,
  classNo: PropTypes.string,
  teacher: PropTypes.string,
  id: PropTypes.string
};

export default ActivitiesListItem;
