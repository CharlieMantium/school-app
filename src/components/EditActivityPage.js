import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import ActivityForm from './ActivityForm';
import { editActivity, removeActivity } from '../store/activities/actions';
import { getEditedActivity } from '../store/activities/selectors';
import { ACTIVITY_PLAN_ROUTE } from '../constants/routes';
import activityPropTypeShape from '../prop-types/activity';
import historyPushPropTypeShape from '../prop-types/history';

export const EditActivityPage = props => (
  <div>
    <ActivityForm
      activity={props.activity}
      onSubmit={activity => {
        props.onEditActivity(props.activity.id, activity);
        props.history.push(ACTIVITY_PLAN_ROUTE);
      }}
    />
    <button
      onClick={() => {
        props.onRemoveActivity(props.activity.id);
        props.history.push(ACTIVITY_PLAN_ROUTE);
      }}
      type="submit"
    >
      Remove
    </button>
  </div>
);

EditActivityPage.propTypes = {
  activity: PropTypes.shape(activityPropTypeShape),
  history: PropTypes.shape(historyPushPropTypeShape).isRequired,
  onEditActivity: PropTypes.func.isRequired,
  onRemoveActivity: PropTypes.func.isRequired,
};

EditActivityPage.defaultProps = {
  activity: {},
};

const mapStateToProps = (state, props) => ({
  activity: getEditedActivity(state, props),
});

const mapDispatchToProps = {
  onEditActivity: editActivity,
  onRemoveActivity: removeActivity,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(EditActivityPage);
