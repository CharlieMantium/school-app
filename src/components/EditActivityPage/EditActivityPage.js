import React, { useCallback, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { get } from 'lodash';
import styled from 'styled-components';
import Loader from 'react-loader';
import database from '../../firebase/firebase';

import ActivityForm from '../ActivityForm/ActivityForm';
import { startEditActivity, startRemoveActivity } from '../../store/activities/actions';
import { getEditedActivity } from '../../store/activities/selectors';
import { ACTIVITY_PLAN_ROUTE } from '../../constants/routes';
import activityPropTypeShape from '../../prop-types/activity';
import historyPushPropTypeShape from '../../prop-types/history';
import { generateActivitiesItemsPath } from '../../helpers/paths';

const Button = styled.button`
  color: red;
`;

const EditActivityPage = ({
  activity,
  history,
  onStartEditActivity,
  onStartRemoveActivity,
  match: {
    params: { id },
  },
}) => {
  const [isIdLoaded, changeIdLoaded] = useState(false);
  const [editedActivity, setEditedActivity] = useState({ ...activity, id });

  const activityId = get(editedActivity, 'id', id);

  const asyncEditActivity = useCallback(
    async activitySubmited => {
      try {
        await onStartEditActivity(activityId, activitySubmited);
        history.push(ACTIVITY_PLAN_ROUTE);
      } catch (error) {
        // eslint-disable-next-line no-console
        console.log(`Something went wrong durring activity editing: ${error}`);
      }
    },
    [activityId],
  );

  const asyncRemoveActivity = useCallback(async () => {
    try {
      await onStartRemoveActivity(activityId);
      history.push(ACTIVITY_PLAN_ROUTE);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log(`Something went wrong durring activity removal: ${error}`);
    }
  }, [activityId]);

  useEffect(() => {
    (async () => {
      const snapshot = await database.ref(generateActivitiesItemsPath(id)).once('value');
      const fetchedActivity = snapshot.val();
      setEditedActivity({ ...fetchedActivity, id: snapshot.key });
      changeIdLoaded(!isIdLoaded);
    })();
  }, [id]);

  return (
    <div>
      <Loader loaded={isIdLoaded} data-test="loader">
        <ActivityForm
          activity={editedActivity}
          onSubmit={activitySubmited => asyncEditActivity(activitySubmited)}
          data-test="form"
        />
      </Loader>

      <Button
        onClick={() => asyncRemoveActivity(activityId)}
        type="submit"
        data-test="button-remove"
      >
        Remove
      </Button>
    </div>
  );
};

EditActivityPage.propTypes = {
  activity: PropTypes.shape(activityPropTypeShape),
  history: PropTypes.shape(historyPushPropTypeShape).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }),
  }).isRequired,
  onStartEditActivity: PropTypes.func.isRequired,
  onStartRemoveActivity: PropTypes.func.isRequired,
};

EditActivityPage.defaultProps = {
  activity: null,
};

const mapStateToProps = (state, props) => ({
  activity: getEditedActivity(state, props),
});

const mapDispatchToProps = {
  onStartEditActivity: startEditActivity,
  onStartRemoveActivity: startRemoveActivity,
};

export { EditActivityPage as EditActivityPageUnwrapped };
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(EditActivityPage);

// TODO: I recommend you to think about react-intl package.
// This is not a priority, but it would be great knowledge for future ;)
// spend some time for internatiolization in short future.
