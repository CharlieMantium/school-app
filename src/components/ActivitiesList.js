import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import ActivitiesListItem from './ActivitiesListItem';
import selectActivities from '../store/activities/selectors/activities';

const ActivitiesList = ({ activities }) => (
  <div>
    <h1>Activities List</h1>
    {activities.map((activity) => (
      <ActivitiesListItem {...activity} key={activity.id}/>
    ))}
  </div>
);

ActivitiesList.propTypes = {
  activities: PropTypes.array
};

const mapStateToProps = (state) => ({
  activities: selectActivities(state.activities, state.filters)
});

export default connect(mapStateToProps)(ActivitiesList);
