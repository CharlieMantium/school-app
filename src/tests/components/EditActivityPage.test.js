import React from 'react';
import { shallow } from 'enzyme';

import { ACTIVITY_PLAN_ROUTE } from '../../constants/routes';
import { EditActivityPage } from '../../components/EditActivityPage';
import testState from '../fixtures/state';

let onEditActivitySpy;
let onRemoveActivitySpy;
let historySpy;
let wrapper;

beforeEach(() => {
  onEditActivitySpy = jest.fn();
  onRemoveActivitySpy = jest.fn();
  historySpy = { push: jest.fn() };
  wrapper = shallow(
    <EditActivityPage
      activity={testState.activities.items[1]}
      onEditActivity={onEditActivitySpy}
      onRemoveActivity={onRemoveActivitySpy}
      history={historySpy}
    />,
  );
});

describe('EditActivityPage', () => {
  it('should render EditActivityPage correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle onSubmit', () => {
    wrapper.find('ActivityForm').prop('onSubmit')(testState.activities.items[1]);
    expect(historySpy.push).toHaveBeenLastCalledWith(ACTIVITY_PLAN_ROUTE);
    expect(onEditActivitySpy).toHaveBeenLastCalledWith(
      testState.activities.items[1].id,
      testState.activities.items[1],
    );
  });

  it('should handle onRemoveActivities', () => {
    wrapper.find('button').simulate('click');
    expect(historySpy.push).toHaveBeenLastCalledWith(ACTIVITY_PLAN_ROUTE);
    expect(onRemoveActivitySpy).toHaveBeenLastCalledWith(testState.activities.items[1].id);
  });
});
