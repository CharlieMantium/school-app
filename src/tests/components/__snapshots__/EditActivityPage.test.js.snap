// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditActivityPage should render EditActivityPage correctly 1`] = `
<div>
  <ActivityForm
    activity={
      Object {
        "classNo": "1",
        "day": "monday",
        "id": "2",
        "name": "polski",
        "room": "hell",
        "teacher": "Malski",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
    type="submit"
  >
    Remove
  </button>
</div>
`;
